cooked-validators/src/Cooked/MockChain/Monad/Staged.hs

```diff
@@ -83,10 +105,17 @@ instance Monad StagedMockChain where
   (Return x) >>= f = f x
   (Instr i m) >>= f = Instr i (m >=> f)
 
+instance MonadFail StagedMockChain where
+  fail = singleton . Fail
+
+instance Alternative StagedMockChain where
+  empty = singleton MZero
+  a <|> b = singleton $ MPlus a b
+
 -- | Interprets a single operation in the direct 'MockChainT' monad. We need to
 -- pass around the modalities since some operations refer back to 'StagedMockChain',
 -- and that needs to be interpreted taking into account the existing modalities.
-interpretOp :: MockChainOp a -> InterpMockChainMod a
+interpretOp :: MonadPlus m => MockChainOp a -> InterpMockChainMod m a
```

Comment:
Interesting! I did not know of `Alt`, thank you! I'd be against using `Alt` here however. Firstly because we need `mzero`/`empty` in the `interpretNonDet` function. Secondly, we don't really plan on instantiating `interpretOp` with that many different types to bring in another dependency. 

I declared `interpretOp` for an arbitrary monad plus because I thought of using a rose-tree instead of a list for efficiency. After further consideration, I decided not to "fix" something that wasn't a problem yet, but left the `MonadPlus` in there since the body of the function actually didn't change. :)